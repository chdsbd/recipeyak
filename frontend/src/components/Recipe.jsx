import React from 'react'
import { Link } from 'react-router-dom'

const MetaData = ({
  author = '',
  source = '',
  servings = '',
  time = ''
}) => {
  const isValid = x => x !== '' && x != null

  const _author = isValid(author)
    ? `By ${author} `
    : ''
  const _source = isValid(source)
    ? `from ${source} `
    : ''
  const _servings = isValid(servings)
    ? `creating ${servings} `
    : ''
  const _time = isValid(time)
    ? `in ${time} `
    : ''

  return <p className="">{ _author + _source + _servings + _time }</p>
}

const RecipeViewing = ({
  id,
  name,
  author,
  source,
  servings,
  time,
  ingredients = [],
  steps = [],
  inCart = 0,
  addToCart,
  removeFromCart,
  addingToCart = false,
  removingFromCart = false,
  loading = false,
  error404 = false
}) => {
  if (error404) {
    return <p>404</p>
  }
  if (loading) {
    return <p>Loading...</p>
  }
  return (
    <div className="d-grid grid-template-columns-repeat-12-fr grid-gap-1rem">

      <div className="grid-entire-row d-flex align-center justify-space-between flex-wrap">
        <h1 className="title fs-3rem mb-0">{ name }</h1>
        <div className="d-flex">
          <div>
            <input
              onClick={ () => removeFromCart(id) }
              className={ `button ${removingFromCart ? 'is-loading' : ''}` }
              disabled={ inCart <= 0 }
              type="button"
              value="-"/>
            <span className="tag is-light is-medium cart-count-tag">{ inCart }</span>
            <input
              onClick={ () => addToCart(id) }
              className={ `button is-primary ${addingToCart ? 'is-loading' : ''}` }
              type="button"
              value="+"/>
          </div>
          <Link to={ `/recipes/${id}/edit` } className="button is-link ml-2">Edit</Link>
        </div>
      </div>

      <div className="grid-entire-row">
        <MetaData
          name={name}
          author={author}
          source={source}
          servings={servings}
          time={time}/>
      </div>

      <section className="ingredients-preparation-grid">
        <div>
          <h2 className="title">Ingredients</h2>
          <div className="box">
            <ul>
              {
                ingredients.map(({ id, quantity, name, description }) =>
                  <p key={ id } className="listitem-text justify-space-between">
                    { quantity } { name } { description }
                  </p>
                )
              }
            </ul>
          </div>
        </div>

        <div >
          <h2 className="title is-3">Preparation</h2>
          <div className="box">
            <ul>
              {
                steps.map(({ id, text }, i) =>
                  <div key={id}>
                    <label className="label">Step { i + 1}</label>
                    <p className="listitem-text">text</p>
                  </div>

                )
              }
            </ul>
          </div>
        </div>
      </section>
    </div>
  )
}

class Recipe extends React.Component {
  componentWillMount = () => {
    this.props.fetchRecipe(this.props.match.params.id)
  }

  render () {
    return <RecipeViewing
      { ...this.props }
    />
  }
}

export default Recipe
