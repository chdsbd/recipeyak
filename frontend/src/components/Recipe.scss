@import './main.scss';

.flex-center {
  display: flex;
  justify-content: center;
}

.tags > .tag {
  margin-right: 0.35rem;
}

.tags {
  margin-bottom: 0.75rem;
}

$danger: #ff3860;

$border-color: #dbdbdb;

$darkdesertsand: darken($desertsand, 30%);
$desertsand-shadow: transparentize($darkdesertsand, 0.2);

.my-textarea,
.my-input {
  align-items: center;
  border: 1px solid transparent;
  border-radius: 3px;
  box-shadow: none;
  font-size: 1rem;
  height: 2.25em;
  justify-content: flex-start;
  line-height: 1.5;
  padding-bottom: calc(0.375em - 1px);
  padding-left: calc(0.625em - 1px);
  padding-right: calc(0.625em - 1px);
  padding-top: calc(0.375em - 1px);
  position: relative;
  vertical-align: top;
  background-color: white;
  border-color: $border-color;
  color: #363636;
  box-shadow: inset 0 1px 2px rgba(10, 10, 10, 0.1);
  width: 100%;
  outline: none;
  transition: all 0.2s;

  &:focus {
    border-color: $darkdesertsand;
    box-shadow: 0 0 5px 0px $desertsand-shadow;
  }

  &.is-danger {
    border-color: $danger;
  }
}

.my-textarea {
  display: block;
  max-height: 600px;
  max-width: 100%;
  min-height: 120px;
  min-width: 100%;
  padding: 0.625em;
  resize: vertical;
}

.my-box {
  display: block;
  background-color: white;
  border-radius: 5px;
  box-shadow: 0 2px 3px rgba(10, 10, 10, 0.1), 0 0 0 1px rgba(10, 10, 10, 0.1);
  padding: 1.25rem;
}

.my-button {
  align-items: center;
  border: 1px solid transparent;
  border-radius: 3px;
  display: inline-flex;
  font-size: 1rem;
  height: 2.25em;
  justify-content: flex-start;
  line-height: 1.5;
  padding-bottom: calc(0.375em - 1px);
  padding-left: calc(0.625em - 1px);
  padding-right: calc(0.625em - 1px);
  padding-top: calc(0.375em - 1px);
  position: relative;
  vertical-align: top;
  user-select: none;
  background-color: white;
  border-color: $border-color;
  color: #363636;
  cursor: pointer;
  justify-content: center;
  padding-left: 0.75em;
  padding-right: 0.75em;
  text-align: center;
  white-space: nowrap;
  outline: none;
  transition: all 0.2s;

  &:active {
    transform: translateY(1px);
    border-color: $darkdesertsand;
    box-shadow: inset 0 1px 2px rgba(10, 10, 10, 0.2);
  }

  &:focus {
    border-color: $darkdesertsand;
    box-shadow: 0 0 5px 0px $desertsand-shadow;
  }

  &:hover {
    border-color: $darkdesertsand;
  }

  &.is-large {
    font-size: 1.5rem;
  }

  &.is-primary {
    $primary-color: darken($desertsand, 25%);
    background-color: $primary-color;
    border-color: transparent;
    color: #fff;

    &:active {
      background-color: darken($primary-color, 10%) !important;
      box-shadow: inset 0 1px 2px rgba(10, 10, 10, 0.2);
    }

    &:hover:not([disabled]) {
      background-color: darken($primary-color, 5%);
    }

    &:focus {
      box-shadow: 0 0 5px $desertsand-shadow;
    }
  }

  &[disabled] {
    box-shadow: none;
    opacity: 0.5;
    cursor: not-allowed;
  }


}

.c-danger {
  color: $danger;
}
