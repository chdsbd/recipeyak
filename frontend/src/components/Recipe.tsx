import React from "react"
import { connect } from "react-redux"
import { Helmet } from "./Helmet"

import NoMatch from "./NoMatch"
import Loader from "./Loader"
import AddStep from "./AddStep"
import AddIngredient, { IIngredientBasic } from "./AddIngredient"
import StepContainer from "./StepContainer"
import Ingredient from "./Ingredient"
import RecipeTitle from "./RecipeTitle"

import {
  addingRecipeIngredient,
  addingRecipeStep,
  fetchRecipe,
  deletingRecipe,
  updateRecipe,
  deletingIngredient,
  updatingIngredient,
  Dispatch
} from "../store/actions"
import { RootState } from "../store/store"
import { RouteComponentProps } from "react-router"
import { IRecipe, IStep, IIngredient } from "../store/reducers/recipes"
import { IRecipeBasic } from "./RecipeTitle"

type RouteProps = RouteComponentProps<{ id: string }>

const mapStateToProps = (state: RootState, props: RouteProps) => {
  const id = props.match.params.id
  const recipe = state.recipes[id] ? state.recipes[id] : { loading: true }
  return recipe
}

const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchRecipe: (id: number) => dispatch(fetchRecipe(id)),
  addIngredient: (recipeID: IRecipe["id"], ingredient: IIngredient) =>
    dispatch(addingRecipeIngredient(recipeID, ingredient)),
  addStep: (id: IStep["id"], step: IStep) =>
    dispatch(addingRecipeStep(id, step)),
  update: (id: IRecipe["id"], data: IRecipe) =>
    dispatch(updateRecipe(id, data)),
  remove: (id: IRecipe["id"]) => dispatch(deletingRecipe(id)),
  updateIngredient: (
    recipeID: IRecipe["id"],
    ingredientID: IIngredient["id"],
    content: IIngredient
  ) => dispatch(updatingIngredient(recipeID, ingredientID, content)),
  removeIngredient: (
    recipeID: IRecipe["id"],
    ingredientID: IIngredient["id"]
  ) => dispatch(deletingIngredient(recipeID, ingredientID))
})

interface IRecipeProps extends RouteProps {
  readonly id: IRecipe["id"]
  readonly name: IRecipe["name"]
  readonly author: IRecipe["author"]
  readonly source: IRecipe["source"]
  readonly servings: IRecipe["servings"]
  readonly time: IRecipe["time"]
  readonly ingredients: IIngredient[]
  readonly steps: IStep[]
  readonly loading: boolean
  readonly error404: boolean
  readonly owner: IRecipe["owner"]
  readonly update: (id: IRecipe["id"], recipe: IRecipeBasic) => Promise<void>
  readonly remove: () => void
  readonly deleting: boolean
  readonly updating: boolean
  readonly addingStepToRecipe: boolean
  readonly last_scheduled: string
  readonly fetchRecipe: (id: IRecipe["id"]) => void
  readonly addIngredient: (
    id: number,
    { quantity, name, description }: IIngredientBasic
  ) => void
  readonly updateIngredient: (
    recipeID: IRecipe["id"],
    ingredientID: IIngredient["id"],
    content: IIngredient
  ) => void
  readonly removeIngredient: (
    recipeID: IRecipe["id"],
    ingredientID: IIngredient["id"]
  ) => void
  readonly addStep: () => void
}

interface IRecipeState {
  readonly show: boolean
  readonly addStep: boolean
  readonly addIngredient: boolean
}

class Recipe extends React.Component<IRecipeProps, IRecipeState> {
  state: IRecipeState = {
    show: false,
    addStep: false,
    addIngredient: false
  }

  componentWillMount() {
    this.props.fetchRecipe(parseInt(this.props.match.params.id, 10))
  }

  render() {
    const { id, name, ingredients, steps, loading, error404 } = this.props
    if (error404) {
      return <NoMatch />
    }
    if (loading) {
      return (
        <section className="d-flex justify-content-center">
          <Loader />
        </section>
      )
    }
    return (
      <div className="d-grid grid-gap-2">
        <Helmet title={name} />

        <RecipeTitle
          id={this.props.id}
          name={this.props.name}
          author={this.props.author}
          source={this.props.source}
          servings={this.props.servings}
          time={this.props.time}
          owner={this.props.owner}
          lastScheduled={this.props.last_scheduled}
          update={this.props.update}
          updating={this.props.updating}
          remove={this.props.remove}
          deleting={this.props.deleting}
        />
        <section className="ingredients-preparation-grid">
          <div>
            <h2 className="title is-3 mb-1 font-family-title bold">
              Ingredients
            </h2>
            <ul>
              {ingredients.map(ingre => (
                <Ingredient
                  key={ingre.id}
                  recipeID={this.props.id}
                  id={ingre.id}
                  quantity={ingre.quantity}
                  name={ingre.name}
                  update={(ingredient: IIngredient) =>
                    this.props.updateIngredient(
                      this.props.id,
                      ingre.id,
                      ingredient
                    )
                  }
                  remove={() =>
                    this.props.removeIngredient(this.props.id, ingre.id)
                  }
                  updating={ingre.updating}
                  removing={ingre.removing}
                  description={ingre.description}
                  optional={ingre.optional}
                />
              ))}
            </ul>
            {this.state.addIngredient ? (
              <AddIngredient
                id={id}
                autoFocus
                loading={false}
                addIngredient={this.props.addIngredient}
                onCancel={() => this.setState({ addIngredient: false })}
              />
            ) : (
              <a
                className="text-muted"
                onClick={() => this.setState({ addIngredient: true })}>
                add
              </a>
            )}
          </div>

          <div>
            <h2 className="title is-3 mb-1 font-family-title bold">
              Preparation
            </h2>
            <StepContainer steps={steps} recipeID={id} />
            {this.state.addStep ? (
              <AddStep
                id={id}
                index={steps.length + 1}
                autoFocus
                addStep={this.props.addStep}
                onCancel={() => this.setState({ addStep: false })}
                loading={this.props.addingStepToRecipe}
              />
            ) : (
              <a
                className="text-muted"
                onClick={() => this.setState({ addStep: true })}>
                add
              </a>
            )}
          </div>
        </section>
      </div>
    )
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Recipe)
