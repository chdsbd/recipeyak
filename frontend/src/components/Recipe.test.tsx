import React from "react"
import { match as Match } from "react-router"
import { Location } from "history"
import renderer, { act } from "react-test-renderer"
import { mount } from "enzyme"
import { emptyStore as store, history } from "@/store/store"
import { Recipe } from "@/components/Recipe"
import { IRecipe, fetchRecipe } from "@/store/reducers/recipes"
import {
  baseRecipe,
  baseIngredient,
  baseStep
} from "@/store/reducers/recipes.test"
import { TestProvider } from "@/testUtils"
import { RecipeTimeline } from "@/components/RecipeTimeline"

describe("<Recipe/>", () => {
  const match: Match<{ id: string }> = {
    path: "/recipes/:id(\\d+)(.*)",
    url: "/recipes/98-apple-crisp",
    isExact: true,
    params: {
      id: "98"
    }
  }
  const location: Location = {
    pathname: "/recipes/98-apple-crisp",
    search: "?timeline=1",
    hash: "",
    key: "u3gfv7",
    state: null
  }

  it("renders without failure", () => {
    mount(
      <TestProvider>
        <Recipe history={history} match={match} location={location} />
      </TestProvider>
    )
  })

  describe("snaps", () => {
    test("recipe", () => {
      const recipe: IRecipe = {
        ...baseRecipe,
        id: 98,
        name: "Apple Crisp",
        ingredients: [baseIngredient],
        steps: [baseStep]
      }

      const actions = [
        fetchRecipe.failure({ id: recipe.id, error404: true }),
        fetchRecipe.success(recipe)
      ]

      actions.forEach(action => {
        act(() => {
          store.dispatch(action)
        })

        const tree = renderer
          .create(
            <TestProvider>
              <Recipe
                history={history}
                match={match}
                location={{ ...location, search: "?timeline=1" }}
              />
            </TestProvider>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })
    })
    test("timeline", () => {
      const timelineTree = renderer
        .create(
          <TestProvider>
            <RecipeTimeline createdAt="1776-1-1" recipeId={10} />
          </TestProvider>
        )
        .toJSON()
      expect(timelineTree).toMatchSnapshot()
    })
  })
})
