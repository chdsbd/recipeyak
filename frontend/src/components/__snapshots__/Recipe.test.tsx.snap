// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipe/> snaps recipe 1`] = `
<section>
  <h1
    className="fs-10rem bold text-center"
  >
    404
  </h1>
  <p
    className="fs-3rem text-center"
  >
    Nothing here ðŸŒµ
  </p>
</section>
`;

exports[`<Recipe/> snaps recipe 2`] = `
<section
  className="d-flex justify-content-center"
>
  <div
    className="d-flex justify-content-center "
  >
    <div
      className="ball-grid-pulse grid-entire-row justify-self-center"
    >
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</section>
`;

exports[`<Recipe/> snaps recipe all states 1`] = `
.c0 {
  position: relative;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  padding: 0.25rem 0.5rem;
  font-weight: 400;
  color: #212529;
  white-space: nowrap;
  text-align: left;
  cursor: pointer;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  background-color: transparent;
  border: none;
}

.c2:hover {
  color: #16181b;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:active {
  background-color: rgba(255,114,71,0.8);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  padding: 0.25rem 0.5rem;
  font-weight: 400;
  color: #212529;
  white-space: nowrap;
  text-align: left;
  cursor: pointer;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  background-color: transparent;
  border: none;
}

.c3:hover {
  color: #16181b;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3:active {
  background-color: rgba(255,114,71,0.8);
}

.c1 {
  position: absolute;
  left: auto;
  right: 0;
  z-index: 1000;
  padding: 0.5rem;
  margin: 0.125rem 0 0;
  font-size: 1rem;
  white-space: nowrap;
  background-color: #f9f9f9;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 0.25rem;
  display: none;
}

.c9 {
  font-weight: 500;
  padding: 0.5rem;
  margin-bottom: 0.5rem;
  border-radius: 5px;
  border: 1px solid lightgray;
}

.c6 {
  max-width: 600px;
}

.c7 {
  list-style: none;
  margin-left: 0.5rem;
  padding-left: 0.5rem;
  border-left: 3px solid;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.c8 {
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

.c4 {
  font-weight: bold;
  margin-right: 0.5rem;
}

.c4:hover {
  border-bottom: 2px solid;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 {
  font-weight: bold;
  margin-right: 0.5rem;
  border-bottom: 2px solid;
}

.c5:hover {
  border-bottom: 2px solid;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="d-grid grid-gap-2"
>
  <div>
    <div
      className="grid-entire-row d-flex justify-space-between p-rel"
    >
      <div
        className="d-flex align-items-center"
      >
        <h1
          className="title fs-2rem mb-0 mb-1 cursor-pointer"
          onClick={[Function]}
          title="click to edit"
        >
          Apple Crisp
        </h1>
      </div>
      <div
        className="c0"
      >
        <button
          className="my-button is-small"
          onClick={[Function]}
        >
          Options 
          <svg
            className="fill-text-color"
            style={
              Object {
                "width": "1.5em",
              }
            }
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M14.1711599,9.3535 L9.99925636,13.529 L5.82735283,9.3535 C5.51262415,9.0385 5.73543207,8.5 6.18054835,8.5 L13.8179644,8.5 C14.2630807,8.5 14.4858886,9.0385 14.1711599,9.3535"
                fill="inherit"
              />
            </g>
          </svg>
        </button>
        <div
          className="c1"
        >
          <a
            className="c2"
            href="/schedule/recipes?recipeId=98"
            onClick={[Function]}
          >
            Schedule
          </a>
          <button
            className="c3"
            onClick={[Function]}
          >
            <span>
              Duplicate
            </span>
          </button>
          <button
            className="c3"
            onClick={[Function]}
          >
            Copy Ingredients to Clipboard
          </button>
          <a
            className="c2"
            href="/recipes/98.yaml"
            onClick={[Function]}
          >
            Export as YAML
          </a>
          <a
            className="c2"
            href="/recipes/98.json"
            onClick={[Function]}
          >
            Export as JSON
          </a>
          <button
            className="c3"
            onClick={[Function]}
          >
            Archive Recipe
          </button>
          <button
            className="c3"
            onClick={[Function]}
          >
            Delete Recipe
          </button>
        </div>
      </div>
    </div>
    <div
      className="grid-entire-row"
    >
      <div
        className="break-word"
      >
        <span>
          By 
          <b
            className="cursor-pointer"
            onClick={[Function]}
            title="click  to edit"
          >
            bar
          </b>
           
          from
           
          <b
            className="cursor-pointer"
            onClick={undefined}
            title="click  to edit"
          >
            <a
              href="foo.com"
            >
              foo.com
            </a>
          </b>
           
          creating 
          <b
            className="cursor-pointer"
            onClick={[Function]}
            title="click  to edit"
          >
            12 servings
          </b>
           
          in 
          <b
            className="cursor-pointer"
            onClick={[Function]}
            title="click  to edit"
          >
            1776
          </b>
           
        </span>
        <span
          className="fw-500 p-rel"
        >
          <span
            className="cursor-pointer"
            onClick={[Function]}
            title="click to edit"
          >
            via
          </span>
           
          <a
            href="/t/1-you"
            onClick={[Function]}
          >
            you
          </a>
          <div
            className=" d-none"
          >
            <div
              className="p-abs"
            >
              <div
                className="dropdown-content"
              >
                <div
                  className="text-center"
                >
                  Teams
                </div>
                <hr
                  className="dropdown-divider mt-1 mb-1"
                />
                <div
                  className="select is-normal is-multiple max-height-25vh overflow-y-scroll w-100"
                >
                  <select
                    className="my-select"
                    multiple={true}
                    onChange={[Function]}
                    value={Array []}
                  >
                    <option
                      className="text-small fw-500"
                      value="null-user"
                    >
                      personal
                    </option>
                  </select>
                </div>
                <hr
                  className="dropdown-divider"
                />
                <div
                  className="d-flex justify-space-between ml-2 mr-2"
                >
                  <button
                    className="my-button undefined is-link is-small"
                    onClick={[Function]}
                  >
                    cancel
                  </button>
                  <div
                    className="d-flex justify-space-between"
                  >
                    <button
                      className="my-button mr-1 is-secondary is-small"
                      disabled={true}
                      onClick={[Function]}
                    >
                      move
                    </button>
                    <button
                      className="my-button undefined is-primary is-small"
                      disabled={true}
                      onClick={[Function]}
                    >
                      copy
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <a
      className="c4"
      href="/recipes/98-apple-crisp"
      isActive={false}
      onClick={[Function]}
    >
      Detail
    </a>
    <a
      className="c5"
      href="/recipes/98-apple-crisp?timeline=1"
      isActive={true}
      onClick={[Function]}
    >
      Timeline
    </a>
  </div>
  <div
    className="c6"
  >
    <h2
      className="bold fs-6 mb-0"
    >
      Timeline
    </h2>
    <ol
      className="c7"
    >
      <div
        className="c8"
      >
        Loading timeline..
      </div>
      <li
        className="c9"
        type="created"
      >
        ðŸŽ‰ Recipe created on 
        <time
          dateTime="1776-1-1"
          title="Jan 1, 1776"
        >
          Jan 1, 1776
        </time>
      </li>
    </ol>
  </div>
</div>
`;

exports[`<Recipe/> snaps recipe all states 2`] = `
.c6 a {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6 a:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6 ul {
  list-style-type: disc;
  padding-left: 1.5rem;
}

.c6 ol {
  padding-left: 1.5rem;
}

.c6 blockquote {
  padding-left: 0.25rem;
  border-left: 5px solid lightgray;
}

.c6 blockquote > p {
  margin-bottom: 0rem;
}

.c6 blockquote:last-of-type > p {
  margin-bottom: 0.5rem;
}

.c6 p {
  margin-bottom: 0.5rem;
}

.c0 {
  position: relative;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  padding: 0.25rem 0.5rem;
  font-weight: 400;
  color: #212529;
  white-space: nowrap;
  text-align: left;
  cursor: pointer;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  background-color: transparent;
  border: none;
}

.c2:hover {
  color: #16181b;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:active {
  background-color: rgba(255,114,71,0.8);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  padding: 0.25rem 0.5rem;
  font-weight: 400;
  color: #212529;
  white-space: nowrap;
  text-align: left;
  cursor: pointer;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  background-color: transparent;
  border: none;
}

.c3:hover {
  color: #16181b;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3:active {
  background-color: rgba(255,114,71,0.8);
}

.c1 {
  position: absolute;
  left: auto;
  right: 0;
  z-index: 1000;
  padding: 0.5rem;
  margin: 0.125rem 0 0;
  font-size: 1rem;
  white-space: nowrap;
  background-color: #f9f9f9;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 0.25rem;
  display: none;
}

.c5 {
  font-weight: bold;
  margin-right: 0.5rem;
}

.c5:hover {
  border-bottom: 2px solid;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  font-weight: bold;
  margin-right: 0.5rem;
  border-bottom: 2px solid;
}

.c4:hover {
  border-bottom: 2px solid;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="d-grid grid-gap-2"
>
  <div>
    <div
      className="grid-entire-row d-flex justify-space-between p-rel"
    >
      <div
        className="d-flex align-items-center"
      >
        <h1
          className="title fs-2rem mb-0 mb-1 cursor-pointer"
          onClick={[Function]}
          title="click to edit"
        >
          Apple Crisp
        </h1>
      </div>
      <div
        className="c0"
      >
        <button
          className="my-button is-small"
          onClick={[Function]}
        >
          Options 
          <svg
            className="fill-text-color"
            style={
              Object {
                "width": "1.5em",
              }
            }
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M14.1711599,9.3535 L9.99925636,13.529 L5.82735283,9.3535 C5.51262415,9.0385 5.73543207,8.5 6.18054835,8.5 L13.8179644,8.5 C14.2630807,8.5 14.4858886,9.0385 14.1711599,9.3535"
                fill="inherit"
              />
            </g>
          </svg>
        </button>
        <div
          className="c1"
        >
          <a
            className="c2"
            href="/schedule/recipes?recipeId=98"
            onClick={[Function]}
          >
            Schedule
          </a>
          <button
            className="c3"
            onClick={[Function]}
          >
            <span>
              Duplicate
            </span>
          </button>
          <button
            className="c3"
            onClick={[Function]}
          >
            Copy Ingredients to Clipboard
          </button>
          <a
            className="c2"
            href="/recipes/98.yaml"
            onClick={[Function]}
          >
            Export as YAML
          </a>
          <a
            className="c2"
            href="/recipes/98.json"
            onClick={[Function]}
          >
            Export as JSON
          </a>
          <button
            className="c3"
            onClick={[Function]}
          >
            Archive Recipe
          </button>
          <button
            className="c3"
            onClick={[Function]}
          >
            Delete Recipe
          </button>
        </div>
      </div>
    </div>
    <div
      className="grid-entire-row"
    >
      <div
        className="break-word"
      >
        <span>
          By 
          <b
            className="cursor-pointer"
            onClick={[Function]}
            title="click  to edit"
          >
            bar
          </b>
           
          from
           
          <b
            className="cursor-pointer"
            onClick={undefined}
            title="click  to edit"
          >
            <a
              href="foo.com"
            >
              foo.com
            </a>
          </b>
           
          creating 
          <b
            className="cursor-pointer"
            onClick={[Function]}
            title="click  to edit"
          >
            12 servings
          </b>
           
          in 
          <b
            className="cursor-pointer"
            onClick={[Function]}
            title="click  to edit"
          >
            1776
          </b>
           
        </span>
        <span
          className="fw-500 p-rel"
        >
          <span
            className="cursor-pointer"
            onClick={[Function]}
            title="click to edit"
          >
            via
          </span>
           
          <a
            href="/t/1-you"
            onClick={[Function]}
          >
            you
          </a>
          <div
            className=" d-none"
          >
            <div
              className="p-abs"
            >
              <div
                className="dropdown-content"
              >
                <div
                  className="text-center"
                >
                  Teams
                </div>
                <hr
                  className="dropdown-divider mt-1 mb-1"
                />
                <div
                  className="select is-normal is-multiple max-height-25vh overflow-y-scroll w-100"
                >
                  <select
                    className="my-select"
                    multiple={true}
                    onChange={[Function]}
                    value={Array []}
                  >
                    <option
                      className="text-small fw-500"
                      value="null-user"
                    >
                      personal
                    </option>
                  </select>
                </div>
                <hr
                  className="dropdown-divider"
                />
                <div
                  className="d-flex justify-space-between ml-2 mr-2"
                >
                  <button
                    className="my-button undefined is-link is-small"
                    onClick={[Function]}
                  >
                    cancel
                  </button>
                  <div
                    className="d-flex justify-space-between"
                  >
                    <button
                      className="my-button mr-1 is-secondary is-small"
                      disabled={true}
                      onClick={[Function]}
                    >
                      move
                    </button>
                    <button
                      className="my-button undefined is-primary is-small"
                      disabled={true}
                      onClick={[Function]}
                    >
                      copy
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <a
      className="c4"
      href="/recipes/98-apple-crisp"
      isActive={true}
      onClick={[Function]}
    >
      Detail
    </a>
    <a
      className="c5"
      href="/recipes/98-apple-crisp?timeline=1"
      isActive={false}
      onClick={[Function]}
    >
      Timeline
    </a>
  </div>
  <section
    className="ingredients-preparation-grid"
  >
    <div>
      <h2
        className="bold fs-6 mb-0"
      >
        Ingredients
      </h2>
      <ul>
        <li
          className="bg-white"
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <section
            className="cursor-pointer"
            onClick={[Function]}
            title="click to edit"
          >
            <p
              className="listitem-text justify-space-between"
            >
              1 cup
               
              foo
              , chopped
               
              
            </p>
          </section>
        </li>
      </ul>
      <a
        className="text-muted"
        onClick={[Function]}
      >
        add
      </a>
    </div>
    <div>
      <h2
        className="bold fs-6 mb-0"
      >
        Preparation
      </h2>
      <div
        className="mb-2"
        style={
          Object {
            "backgroundColor": "white",
            "opacity": 1,
          }
        }
      >
        <label
          className="better-label"
          style={
            Object {
              "cursor": "move",
            }
          }
        >
          Step 
          1
        </label>
        <div>
          <section
            className="cursor-pointer"
            onClick={[Function]}
            title="click to edit"
          >
            <div
              className="c6"
              onClick={undefined}
              title={undefined}
            >
              <p>
                foo
              </p>
            </div>
          </section>
        </div>
      </div>
      <a
        className="text-muted"
        onClick={[Function]}
      >
        add
      </a>
    </div>
    <div />
    <div>
      <h2
        className="bold fs-6 mb-0"
      >
        Notes
      </h2>
      <div>
        <textarea
          className="textarea"
          id="new_note_textarea"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Add a note..."
          rows={1}
          style={
            Object {
              "height": 0,
            }
          }
          value=""
        />
      </div>
      <hr />
    </div>
  </section>
</div>
`;

exports[`<Recipe/> snaps timeline 1`] = `
.c3 {
  font-weight: 500;
  padding: 0.5rem;
  margin-bottom: 0.5rem;
  border-radius: 5px;
  border: 1px solid lightgray;
}

.c0 {
  max-width: 600px;
}

.c1 {
  list-style: none;
  margin-left: 0.5rem;
  padding-left: 0.5rem;
  border-left: 3px solid;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.c2 {
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

<div
  className="c0"
>
  <h2
    className="bold fs-6 mb-0"
  >
    Timeline
  </h2>
  <ol
    className="c1"
  >
    <div
      className="c2"
    >
      Loading timeline..
    </div>
    <li
      className="c3"
      type="created"
    >
      ðŸŽ‰ Recipe created on 
      <time
        dateTime="1776-1-1"
        title="Jan 1, 1776"
      >
        Jan 1, 1776
      </time>
    </li>
  </ol>
</div>
`;

exports[`<Recipe/> snaps useRecipe hook: Refetching 1`] = `
Object {
  "data": Object {
    "archived_at": null,
    "author": "bar",
    "created": "1776-1-1",
    "edits": Array [],
    "id": 98,
    "ingredients": Array [
      Object {
        "description": "chopped",
        "id": 1,
        "name": "foo",
        "optional": false,
        "position": 12.3,
        "quantity": "1 cup",
      },
    ],
    "last_scheduled": "1779",
    "modified": "1778",
    "name": "Apple Crisp",
    "notes": Array [],
    "owner": Object {
      "id": 1,
      "type": "user",
    },
    "sections": Array [],
    "servings": "12 servings",
    "source": "foo.com",
    "steps": Array [
      Object {
        "id": 1,
        "position": 10,
        "text": "foo",
      },
    ],
    "team": 1,
    "time": "1776",
  },
  "kind": "Refetching",
}
`;
